feat(stats): Purge query digest based on last seen

wazir-ahmed opened the pull request:
- Add handler for purge query digest command
- Handle last_seen argument in purge query digest operations
- Make purge logic common for both sync & async operations

Closes #4543.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added time-based filtering for query digest purging operations, allowing selective purging of digests based on last-seen timestamp.
  * Introduced time conversion utility functions for monotonic to realtime clock conversions.

* **Refactor**
  * Simplified query digest purge method signatures; removed message buffer parameters in favor of time-based filtering.
  * Enhanced PURGE command parsing to support timestamp-based digest filtering with server-type discrimination.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wazir-ahmed committed:
feat(stats): Purge query digest based on last seen

- Add handler for purge query digest command
- Handle last_seen argument in purge query digest operations
- Make purge logic common for both sync & async operations

Signed-off-by: Wazir Ahmed <wazirahmedf@gmail.com>

wazir-ahmed commented:
@renecannao PTAL

renecannao commented:
Add to whitelist

renecannao commented:
Can one of the admins verify this patch?

rahim-kanji commented on lib/Admin_Handler.cpp:
```diff
@@ -322,6 +342,38 @@ bool FlushCommandWrapper(S* sess, const string& modname, char *query_no_space, i
+std::tuple<bool, enum SERVER_TYPE, time_t>&& parse_command_purge_query_digests(char *query, int query_len) {
+	auto ret = new std::tuple<bool, enum SERVER_TYPE, time_t>(false, SERVER_TYPE_MYSQL, 0);
+
+	const char *prefix = match_command_prefix(CMD_PREFIX_PURGE_QUERY_DIGESTS, query, query_len);
```
Currently, `ret `is allocated via `new`, but it's never deleted. Even though you `std::move(*ret)`, the heap memory itself is leaked each time this function is called.

Suggestion: You don't need dynamic allocation here. Just declare std::tuple on the stack and return it by value. 
`std::tuple<bool, enum SERVER_TYPE, time_t> ret(false, SERVER_TYPE_MYSQL, 0);`
`return ret;`
_The compiler is capable of applying NRVO for optimization. (For Ref: [RVO/NRVO](https://pvs-studio.com/en/blog/terms/6516/))_

Function declaration: `std::tuple<bool, enum SERVER_TYPE, time_t> parse_command_purge_query_digests(char *query, int query_len);`

wazir-ahmed replied:
Addressed.

rahim-kanji commented on lib/Admin_Handler.cpp:
```diff
@@ -322,6 +342,38 @@ bool FlushCommandWrapper(S* sess, const string& modname, char *query_no_space, i
+		}
+
+		// parse timestamp
+		char *timestamp = strdup(query + strlen(prefix));
```
The timestamp value is not being validated. For instance:
`PURGE TABLE stats.stats_mysql_query_digest TO 123abc;`
The "abc" part is currently ignored by `atoi` (123abc -> 123), but instead of ignoring it, we should return an error to the client indicating that an invalid timestamp value was provided. This will ensure that ProxySQL behaves as expected and avoids any potential unknown issues.

wazir-ahmed replied:
Addressed. Now, the system throws an error - `Invalid timestamp`.

rahim-kanji commented on lib/Admin_Handler.cpp:
```diff
@@ -322,6 +342,38 @@ bool FlushCommandWrapper(S* sess, const string& modname, char *query_no_space, i
+
+		// parse timestamp
+		char *timestamp = strdup(query + strlen(prefix));
+		int last_seen = atoi(trim_spaces_in_place(timestamp));
```
`atoi` is used, risking overflow for large timestamps (e.g., post-2038).

Suggestion: 
* Replace `atoi `with `strtoll`.
* Change `last_seen` datatype to `time_t`

wazir-ahmed replied:
Addressed.

rahim-kanji commented on lib/Admin_Handler.cpp:
```diff
@@ -322,6 +342,38 @@ bool FlushCommandWrapper(S* sess, const string& modname, char *query_no_space, i
+		}
+
+		// set last_seen
+		std::get<2>(*ret) = realtime_to_monotonic_time(last_seen);
```
Suggestion: Instead of relying on tuple indexing with `std::get`, use individual named variables that represent the meaningful names for each of the tuple's elements. This can improve the code's clarity and make it easier to understand and maintain. At the end of the function, return the values as a tuple if needed.
```
bool match = false;
SERVER_TYPE server_type = SERVER_TYPE_MYSQL;
time_t time_stamp = 0;
...
...
return std::make_tuple(match, server_type, time_stamp);
```

wazir-ahmed replied:
Addressed.

rahim-kanji commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
-			delete qds;
-			r++;
+struct purge_query_digests_args {
+	umap_query_digest *digest_umap;
```
Not a problem at the moment, but just a suggestion for future-proofing: it's a good practice to initialize pointers with `nullptr `to avoid potential issues down the line.
```suggestion
	umap_query_digest *digest_umap = nullptr;
```

wazir-ahmed replied:
Done.

rahim-kanji commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
-			r++;
+struct purge_query_digests_args {
+	umap_query_digest *digest_umap;
+	umap_query_digest_text *digest_text_umap;
```
Not a problem at the moment, but just a suggestion for future-proofing: it's a good practice to initialize pointers with `nullptr `to avoid potential issues down the line.
```suggestion
	map_query_digest_text *digest_text_umap = nullptr;
```

wazir-ahmed replied:
Done.

rahim-kanji commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
+	std::function<bool(std::unordered_map<uint64_t, void *>::iterator it)> delete_entry = nullptr;
+
+	// only used for synchronous multi-threaded purge operation
+	int scan_idx;
```
Not a problem at the moment, but just a suggestion for future-proofing: it's a good practice to initialize values to avoid potential issues down the line.
```suggestion
	int scan_idx = -1;
```

wazir-ahmed replied:
Done.

rahim-kanji commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
+		}
+
+		// merge text entries in the aux map with digest_text_umap
+		digest_text_umap.insert(digest_text_umap_aux.begin(), digest_text_umap_aux.end());
```
`digest_rwlock` will remain locked, potentially causing a deadlock, if an exception is thrown.

wazir-ahmed replied:
@rahim-kanji Are you referring to the possibility of `std::bad_alloc` expection from `umap::insert()`?

rahim-kanji commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
-			if ( pthread_create(&args[i].thr, NULL, &purge_query_digests_parallel, &args[i]) != 0 ) {
+
+		for (int i = 0; i < n; i++) {
+			if (pthread_create(&tid[i], NULL, &purge_query_digests_parallel, &args[i]) != 0 ) {
```
There seems to be a potential race condition in `purge_query_digest_entry `(parallel -  multithreaded version). While `digest_rwlock` is being used in main thread, `digest_umap` and `digest_text_umap` are being modified without any additional locks in multiple threads. This can lead to concurrent modifications, which could cause data inconsistencies or race conditions. It might be a good idea to add proper synchronization or locks around these variables to avoid the risk.

wazir-ahmed replied:
Addressed, refer [this](https://github.com/sysown/proxysql/pull/4920#discussion_r2069975502)  comment.

rahim-kanji commented on lib/gen_utils.cpp:
```diff
@@ -350,3 +350,42 @@ const char* escape_string_backslash_spaces(const char* input) {
+	time(&rt_now);
+
+	return ((mt_now - rt_now + rt) * 1000000);
+}
```
add newline

JavierJF commented on lib/Admin_Handler.cpp:
```diff
@@ -288,6 +299,15 @@ bool is_admin_command_or_alias(const std::vector<std::string>& cmds, char *query
 
+const char * match_command_prefix(const std::vector<std::string>& cmd_prefix, char *query, int query_len) {
+	for (auto &prefix : cmd_prefix) {
+		if (query_len >= prefix.length() && !strncasecmp(prefix.c_str(), query, prefix.length())) {
```
This will result in a warning, signed `int` vs `size_t` from `prefix.length()`.

wazir-ahmed replied:
Addressed.

JavierJF commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
+	unsigned long long curtime1 = monotonic_time();
+	bool selective_purge = (last_seen > 0);
+
+	purge_query_digests_args args = {
```
While `designated initializers` are great, and I hope we adopt them in the future, ProxySQL currently compiles with `C++17` and this is a `C++20` feature. It's likely that we will have issues with older compilers, we should stick to regular list initialization for now (until bumping the C++ version).

wazir-ahmed replied:
Addressed.

JavierJF commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
-			args[i].gtu = &digest_text_umap;
+
+	if (parallel && map_size >= DIGEST_STATS_FAST_MINSIZE) { // multi-threaded purge
+		int n = DIGEST_STATS_FAST_THREADS;
```
There is no need to use here variable length arrays, they are not standard C++ but compiler extensions (g++ and clang++). But since `n` is directly from a constant, we can just remove this unnecessary intermediate variable to avoid trouble.

wazir-ahmed replied:
Addressed.

JavierJF commented on lib/Admin_Handler.cpp:
```diff
@@ -322,6 +342,38 @@ bool FlushCommandWrapper(S* sess, const string& modname, char *query_no_space, i
+		}
+
+		// parse timestamp
+		char *timestamp = strdup(query + strlen(prefix));
```
I know that the codebase has inconsistencies regarding the approach for smart pointers, but it's nice to make use of them in new code. Here is a nice candidate for using the abstraction `mf_unique_ptr<char>` that can be seen in other parts of the codebase.

wazir-ahmed replied:
Addressed.

JavierJF commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
+			auto text_it = digest_text_umap->find(it->first);
+			if (text_it != digest_text_umap->end()) {
+				free(text_it->second);
+				digest_text_umap->erase(text_it);
```
As pointed out by @rahim-kanji, in the original design the size of the container wasn't modified during the freeing operation of the elements. This isn't safe to be done in a multi-threaded operation, i.e., the multi-thread freeing can't modify the size of the container, otherwise, as the size of the container is concurrently modified, invalid operations **will take place**, this sequence of commands against the Admin interface allows to test the algorithm under ASAN an trigger the issue:
```
admin> PROXYSQLTEST 1 500;
Query OK, 500000 rows affected (2.28 sec)

admin> PROXYSQLTEST 5;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

The corruption is reported by ASAN:
```
==313907==ERROR: AddressSanitizer: attempting double-free on 0x50b008199bc0 in thread T37:
    #0 0x75646eafc102 in free /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x5661cb31a2a7 in QP_query_digest_stats::~QP_query_digest_stats() /home/javjarfer/Projects/proxysql_dev/lib/QP_query_digest_stats.cpp:85
    #2 0x5661caf26539 in purge_query_digest_entry(purge_query_digests_args*) /home/javjarfer/Projects/proxysql_dev/lib/Query_Processor.cpp:655
    #3 0x5661caf26731 in purge_query_digests_parallel(void*) /home/javjarfer/Projects/proxysql_dev/lib/Query_Processor.cpp:672
    #4 0x75646ea5cec9 in asan_thread_start /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_interceptors.cpp:234
    #5 0x75646daa3709  (/usr/lib/libc.so.6+0x95709) (BuildId: 0b707b217b15b106c25fe51df3724b25848310c0)
    #6 0x75646db27aab  (/usr/lib/libc.so.6+0x119aab) (BuildId: 0b707b217b15b106c25fe51df3724b25848310c0)

0x50b008199bc0 is located 0 bytes inside of 99-byte region [0x50b008199bc0,0x50b008199c23)
freed by thread T38 here:
    #0 0x75646eafc102 in free /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x5661cb31a2a7 in QP_query_digest_stats::~QP_query_digest_stats() /home/javjarfer/Projects/proxysql_dev/lib/QP_query_digest_stats.cpp:85
    #2 0x5661caf26539 in purge_query_digest_entry(purge_query_digests_args*) /home/javjarfer/Projects/proxysql_dev/lib/Query_Processor.cpp:655
    #3 0x5661caf26731 in purge_query_digests_parallel(void*) /home/javjarfer/Projects/proxysql_dev/lib/Query_Processor.cpp:672
    #4 0x75646ea5cec9 in asan_thread_start /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_interceptors.cpp:234
    #5 0x75646daa3709  (/usr/lib/libc.so.6+0x95709) (BuildId: 0b707b217b15b106c25fe51df3724b25848310c0)

previously allocated by thread T34 here:
    #0 0x75646eafd721 in malloc /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x75646ea771af in strndup /usr/src/debug/gcc/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:416
    #2 0x5661cb319be6 in QP_query_digest_stats::QP_query_digest_stats(char const*, char const*, unsigned long, char const*, int, char const*, int) /home/javjarfer/Projects/proxysql_dev/lib/QP_query_digest_stats.cpp:34
    #3 0x5661caf2a18b in Query_Processor<MySQL_Query_Processor>::update_query_digest(unsigned long, unsigned long, char*, int, MySQL_Connection_userinfo*, unsigned long long, unsigned long long, char const*, unsigned long long, unsigned long long) /home/javjarfer/Projects/proxysql_dev/lib/Query_Processor.cpp:1986
    #4 0x5661cb74dcd5 in ProxySQL_Test___GenerateRandomQueryInDigestTable(int) /home/javjarfer/Projects/proxysql_dev/lib/ProxySQL_Admin_Tests.cpp:139
    #5 0x5661cb34bbeb in void ProxySQL_Admin::ProxySQL_Test_Handler<MySQL_Session>(ProxySQL_Admin*, MySQL_Session*, char*, bool&) /home/javjarfer/Projects/proxysql_dev/lib/ProxySQL_Admin_Tests2.cpp:573
    #6 0x5661cb3d3496 in void admin_session_handler<MySQL_Session>(MySQL_Session*, void*, _PtrSize_t*) /home/javjarfer/Projects/proxysql_dev/lib/Admin_Handler.cpp:2726
    #7 0x5661caea8b43 in MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY___not_mysql(_PtrSize_t&) /home/javjarfer/Projects/proxysql_dev/lib/MySQL_Session.cpp:3452
    #8 0x5661caeb04b3 in MySQL_Session::get_pkts_from_client(bool&, _PtrSize_t&) /home/javjarfer/Projects/proxysql_dev/lib/MySQL_Session.cpp:4208
    #9 0x5661caeb7d19 in MySQL_Session::handler() /home/javjarfer/Projects/proxysql_dev/lib/MySQL_Session.cpp:4821
    #10 0x5661caf89d55 in child_mysql(void*) /home/javjarfer/Projects/proxysql_dev/lib/ProxySQL_Admin.cpp:2045
    #11 0x75646ea5cec9 in asan_thread_start /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_interceptors.cpp:234
    #12 0x75646daa3709  (/usr/lib/libc.so.6+0x95709) (BuildId: 0b707b217b15b106c25fe51df3724b25848310c0)
```

The approach needs to be modified not to concurrently modify the size of the container on multi-threaded purging.

wazir-ahmed replied:
Refer 4fa9a47. I have restricted the multi-threaded approach only to `TRUNCATE` command (when `last_seen == 0`).
- In case of `TRUNCATE`, since this is a case of delete all data, we can do freeing operations in worker threads and restrict erasing items in the container only to parent thread.
- In case of `PURGE`, since this is a case of conditional delete, we cannot separate these operations. At least, I cannot think of a way to efficiently handle this.

I complied the latest code with ASAN, and tested it by triggering `PROXYSQLTEST` commands to check all variations of `purge_query_digests_*` methods. No errors are reported.

JavierJF commented on lib/Query_Processor.cpp:
```diff
@@ -620,123 +620,219 @@ void * get_query_digests_parallel(void *_arg) {
+
+
+void * purge_query_digests_parallel(void *_args) {
+	set_thread_name("PurgeQueryDigest", GloVars.set_thread_name);
```
This will immediately assert when attempting to execute. This name exceeds the maximum name allowed for threads, triggering this assert:
```
	int rc;
	rc = pthread_setname_np(pthread_self(), name);
	assert(!rc);
```
The name should be reverted to the previous one `PurgeQueryDgest`. For executing the test sequence suggested in my concurrent purging comment, this needs to be fixed first.

wazir-ahmed replied:
Addressed.

JavierJF reviewed (commented):
There are issues in the design that needs to be addressed.

wazir-ahmed commented:
@rahim-kanji @JavierJF I have addressed all the review comments (except [this](https://github.com/sysown/proxysql/pull/4920#discussion_r2063173183)). PTAL and let me know if I have missed something.

renecannao committed:
Merge branch 'v3.0' into purge-stats-query-digest

Signed-off-by: René Cannaò <rene@proxysql.com>

renecannao commented:
@wazir-ahmed : we need a TAP test for this (preferred) or modify current a current TAP test .

Given all the concerns , the TAP tests I envision is N threads generating random traffic with different digests ( `SELECT list_of_columns with_random_alias in_random_order FROM tablename LIMIT 1` is perhaps enough ) for some time while an Admin connection is purging at regular interval

renecannao commented:
Bump

renecannao committed:
Merge remote-tracking branch 'origin/v3.0' into pr-4920

renecannao committed:
fix(merge): restore closing brace of realtime_to_monotonic_time lost in v3.0 merge conflict resolution

renecannao committed:
fix(stats): correct digest text lifecycle and stats merge in purge_query_digests

- digest_umap is keyed by digest_total (user+schema+digest+hostgroup+client hash)
  while digest_text_umap is keyed by the plain digest: purge_query_digest_entry
  was looking up texts with the digest_umap key, so no digest text was ever
  freed (leaking every text on TRUNCATE/PURGE when
  query_digests_normalize_digest_text=true). Text release now happens per
  call-site: full purges free the whole text map (restoring pre-PR behavior),
  selective purges garbage-collect texts no longer referenced by any surviving
  entry (a shared text must survive while any user/hostgroup variant of the
  digest survives).
- merging stats collected during the async purge window used
  add_time(aux->min_time, ...), corrupting sum_time/min_time/max_time and
  first_seen; introduce QP_query_digest_stats::merge() doing a field-wise merge.
- the async purge logged the size of the repopulated (new) map instead of the
  purged one, so the timing log would (almost) never print and reported a
  meaningless count; log deleted/original counts instead.
- avoid erasing aux map entries one by one on full async purge, and build the
  referenced-digest set outside the write lock so the critical section stays
  proportional to the purge window.

renecannao committed:
fix(admin): treat pre-boot PURGE timestamps as a no-op, clamp conversion overflow

A 'PURGE stats_mysql_query_digest TO <ts>' with a valid timestamp older than
the system boot was rejected with 'Invalid timestamp' because
realtime_to_monotonic_time() returned a negative value. No digest entry can
predate boot, so such a purge is now a successful no-op (0 rows). Also clamp
realtime_to_monotonic_time() against time_t overflow for very large inputs.

renecannao committed:
test(tap): add comprehensive coverage for PURGE query digest command

Covers: selective purge by last_seen (old digests removed, recent ones and
their stats untouched, digests spanning the cutoff survive with accumulated
count_star), reported affected rows, pre-boot timestamps as successful no-ops,
invalid timestamp errors, all accepted syntax variants including
stats_pgsql_query_digest, digest/text map repopulation after a full purge
(with query_digests_normalize_digest_text=true to exercise the shared text
map), PROXYSQLTEST 4/5 sync purge paths and TRUNCATE, and repeated purges
under concurrent traffic (async purge window reconciliation).

renecannao reviewed (dismissed):
# Review of PR #4920 — `feat(stats): Purge query digest based on last seen`

## Overview

The PR implements #4543: a new admin command `PURGE [TABLE] [stats.]stats_{mysql,pgsql}_query_digest TO <unix-timestamp>` that removes from the in-memory digest map every entry whose `last_seen` is `<=` the given realtime timestamp. To do this it:

- adds `monotonic_time_to_realtime()` / `realtime_to_monotonic_time()` helpers (entries store `last_seen` in monotonic µs, the admin command and the stats tables speak realtime seconds);
- refactors `purge_query_digests_{async,sync}` around a common `purge_query_digest_entry()` helper with an optional match predicate;
- for a **selective async purge**, swaps the whole map out, filters it unlocked, then swaps back and merges entries that accumulated during the purge window;
- removes the `ProxySQL_Test___PurgeDigestTable` indirection (nice cleanup) and wires `PROXYSQLTEST 4/5/6` to the new `last_seen` argument.

The overall direction is good: the command syntax matches the issue, both server types are handled, the refactor removes duplication, and the window-merge idea makes the async selective purge safe in principle. However, the branch **did not compile** as pushed, and the implementation had several real correctness bugs, detailed below. I have merged `v3.0` into the branch and pushed follow-up commits fixing all of them, plus a comprehensive TAP test.

## Blocking issues found (fixed in follow-up commits)

### 1. Branch did not compile — broken merge conflict resolution (`01c141c99`)
The author's earlier "Merge branch 'v3.0'" commit (`60b0d8f88`) dropped the closing `}` of `realtime_to_monotonic_time()` and the opening `/**` of the next doc comment in `lib/gen_utils.cpp`, leaving the file syntactically invalid.

### 2. Digest texts freed with the wrong key — leak + broken TRUNCATE (`e4eda08a7`)
`digest_umap` is keyed by `digest_total` (SpookyHash of username+digest+schemaname+hostgroup+client_addr) while `digest_text_umap` is keyed by the plain `digest`, and **several digest entries can share one text entry**. `purge_query_digest_entry()` did `digest_text_umap->find(it->first)` with the `digest_umap` key, which essentially never matches. Consequences when `query_digests_normalize_digest_text=true`:
- `TRUNCATE stats_mysql_query_digest` (async full purge) leaked **every** digest text;
- the sync single-threaded purge left the text map populated forever (the old `digest_text_umap.erase(...)` was removed);
- the sync parallel purge (`PROXYSQLTEST 5`) leaked all texts (freed nothing, then `clear()`).

Note this is also a design trap, not just a wrong key: even with the "right" key, freeing a text on selective purge is incorrect while another user/hostgroup variant of the same digest survives — `get_digest_text()` would hit `assert(0)` (debug) or use a dangling pointer. The fix frees texts per call-site: full purges free the whole text map (pre-PR behavior), selective purges garbage-collect only texts no longer referenced by any surviving entry.

### 3. Window merge corrupted stats (`e4eda08a7`)
The async selective purge merged entries collected during the purge window with `add_time(aux->min_time, aux->last_seen, ...)`. `add_time()` treats its first argument as a single execution time, so the merged entry got `sum_time += min_time` (instead of `+= sum_time`) and min/max/first_seen handling was wrong. Added a proper field-wise `QP_query_digest_stats::merge()`.

### 4. Purge timing log became dead code (`e4eda08a7`)
`purge_query_digests_async()` read `digest_umap.size()` **after** the map was swapped out, i.e. the size of the freshly repopulating map (≈0), not the purged one. The `proxy_info()` "TRUNCATE ... %llums to remove %lu entries" line would essentially never fire again, and would report a meaningless count if it did. Now logs `deleted/original` based on the pre-swap size.

### 5. Pre-boot timestamps rejected as invalid (`df30112ff`)
`PURGE ... TO <ts>` with a perfectly valid timestamp older than the host's boot time (e.g. a cron job purging "older than 30 days" on a host up for 10 days) produced a negative monotonic value and was rejected with `Invalid timestamp`. No digest can predate boot, so this is now a successful no-op (`0 rows affected`). Also clamped `realtime_to_monotonic_time()` against `time_t` overflow for absurdly large inputs.

## Non-blocking observations (no action taken; opinions welcome)

- **Async vs sync**: #4543 explicitly asked for a *synchronous* purge; the admin command uses the async swap+merge path (`purge_query_digests(true, false, last_seen)`). With the merge fixed this is safe and non-blocking, but it has a visible side effect: while a selective purge runs, a concurrent `SELECT ... FROM stats_mysql_query_digest` sees a (nearly) empty table until the merge-back completes, and a concurrent `TRUNCATE`/reset can race with the merge-back (survivors reappear after the reset). If that is a concern, switching the handler to the sync path (`false, false, last_seen`) is a one-line change — the filtered sync path exists and holds the lock for the whole (fast) scan.
- **`<=` vs `<`**: the issue and the code comment say entries with `last_seen` *less than* `<value>` are deleted; the implementation deletes `last_seen <= value`. Fine either way, but the docs should say `<=`.
- The command requires exactly one space around tokens (`PURGE  TABLE ...` or `TO<ts>` won't match and falls through to a generic SQLite error). Acceptable for an admin command; worth documenting.
- `parallel` purge is silently downgraded to single-threaded when a `last_seen` filter is given (`purge_query_digests_sync`). Only reachable via `PROXYSQLTEST`, so fine, but worth a comment.
- Minor: `monotonic_time_to_realtime()` is called twice per row in `QP_query_digest_stats::get_row()` (first_seen + last_seen), each taking two clock reads; the old code took the clock reads once per row. Negligible, but could be hoisted by the caller if stats dumps of multi-million-entry maps show regression.

## Test coverage

The PR shipped with no tests (the existing `PROXYSQLTEST 4/5/6` paths are exercised nowhere for the new argument). Added `test/tap/tests/test_purge_query_digest-t.cpp` (27 assertions, registered in the same groups as `reg_test_4399-stats_mysql_query_digest-t`):

- selective purge: pre-cutoff digests removed, post-cutoff digests survive, a digest seen on both sides of the cutoff survives with accumulated `count_star`, surviving stats unaltered, affected-rows count correct;
- pre-boot timestamp → successful no-op; `0`, `-1`, `abc`, `123abc` → `Invalid timestamp` error;
- syntax variants (case-insensitive, optional `TABLE`, optional `stats.` prefix) + `stats_pgsql_query_digest`;
- map repopulation after full purge with `query_digests_normalize_digest_text=true` (would trip the `get_digest_text()` assert in a debug build if texts were mishandled);
- `PROXYSQLTEST 1 150` + `PROXYSQLTEST 5` (sync parallel purge over >100k entries), `PROXYSQLTEST 4`, and `TRUNCATE` (async full purge);
- 60 repeated `PURGE ... TO now-1` under concurrent traffic (exercises the async window merge under load).

renecannao committed:
test(tap): use snprintf in purge digest test (SonarCloud S6069)

renecannao committed:
test(unit): adapt query_processor_unit-t to new purge_query_digests signature

The unit test was added to v3.0 after this PR changed purge_query_digests()
from (bool, bool, char**) to (bool, bool, time_t); the v3.0 merge left the
call site on the old signature, breaking the -tap CI builds.

renecannao committed:
Merge branch 'v3.0' into purge-stats-query-digest
